File: android/src/main/kotlin/com/any/wifi_detection/handler/SearchDevicesHandlerImpl.kt
package com.any.wifi_detection.handler

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.any.wifi_detection.async.ScanHostsAsyncTask
import com.any.wifi_detection.network.Wireless
import io.flutter.plugin.common.EventChannel
import java.util.Timer
import java.util.TimerTask


class SearchDevicesHandlerImpl(context: Context) : EventChannel.StreamHandler {

    private val tag = SearchDevicesHandlerImpl::class.java.simpleName


    private val wireless = Wireless(context)

    private val mainHandler = Handler(Looper.getMainLooper())
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        if (!wireless.isEnabled()) {
            events?.error("1", "Wireless is not enabled", null)
            events?.endOfStream()
            return
        }

        if (!wireless.isConnectedWifi()) {
            events?.error("2", "Wireless is not connected", null)
            events?.endOfStream()
            return
        }


        val localIp = wireless.getInternalWifiIpAddress() // 获取内网IP

        val wifiSubnet = wireless.getInternalWifiSubnet() // CIDR 前缀长度


        val timeout = 5000  // 超时时间（毫秒）
        ScanHostsAsyncTask(events, mainHandler)
                .scanHosts(localIp, wifiSubnet, timeout)
        events?.endOfStream() // 结束Stream
    }

    override fun onCancel(arguments: Any?) {
    }
}

File: ios/Classes/handler/SearchDevicesHandlerImpl.swift
//
//  SearchDevicesHandlerImpl.swift
//  wifi_detection
//
//  Created by arthur on 2024/3/4.
//

import Foundation
import Flutter
import Network


class SearchDevicesHandlerImpl: NSObject, FlutterStreamHandler {
    
   let wireless =  Wireless()
    // Handle events on the main thread.
    var timer = Timer()
    // Declare our eventSink, it will be initialized later
//    private var eventSink: FlutterEventSink?
    
    func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
        
        guard wireless.isWiFiConnected()else{
            return FlutterError(code: "2", message: "Wireless is not connected", details: nil)
        }
        let ssid = wireless.getSSID()
        let bssid = wireless.getBSSID()
        let cidr = wireless.getInternalWifiCIDR()
        
        
        print("ssid:\(ssid),bssid:\(bssid),cidr:\(cidr),ip:\(wireless.getInternalWifiIpString())")
        
        let scanHostsAsyncTask = ScanHostsAsyncTask(eventSink: eventSink)
        if let ipv4 = wireless.getInternalWifiIpAddress(),
           let cidr = wireless.getInternalWifiCIDR(){
            print("ipv4: \(ipv4), cidr: \(cidr)")
            scanHostsAsyncTask.scanHosts(ipv4: ipv4, cidr: cidr, timeout: 5000)
        }
        
        
        
        print("onListen......")

        
        self.timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true, block: { _ in
            let dateFormat = DateFormatter()
            dateFormat.dateFormat = "HH:mm:ss"
            let time = dateFormat.string(from: Date())
            eventSink(time)
            eventSink(FlutterEndOfEventStream)
        })
    
        
        
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        
        return nil
    }
}


